### DISCUSSION TITLE
When I dropping Dynamic Graph will be hang up, what can I do to dropped it

### DISCUSSION THREAD

**Discussion by Zorabi** [JanusGraph/janusgraph#4440] [Q&A] [answered]
I have some Dynamic Graph about es and hbase from ConfiguredGraphFactory.When I dropping those, I used `ConfiguradGraphFactory.drop('graphName')`,but it will be hang up until timtout to failed or I used `m.forceCloseInstance('${instanceId}')` to killed instances, the graph will be dropped.

And when I dropped, will be rolling 'graph' table not existed in console.

How to dopped graph safely.

**Comment by CosmoBean** [comment]
I have faced similar issues as well, Since I was using configurationManagementGraph, and CQL-ES config, I deleted the keyspace for the graphname in cql and also the indexed i had created in ES, followed by deleting the mapping from the configuration graph :)

It's not at all clean by any means, if anyone else has a better solution, please let me know

**Comment by FlorianHockmann** [comment]
> it will be hang up until timtout to failed or I used m.forceCloseInstance('${instanceId}') to killed instances

It's unfortunately a known problem that you can sometimes get instances which aren't active any more but JanusGraph doesn't know that they are inactive and will therefore wait for them indefinitely to participate in management operations.

So, if you want to be sure that `ConfiguradGraphFactory.drop('graphName')` will succeed, then the safest approach is probably to first force close all other instances.

We should also definitely describe that more clearly in our docs. #3974 tracks that.

@CosmoBean: Did you also try `ConfiguradGraphFactory.drop('graphName')` and it didn't work or why did you execute these steps manually? `ConfiguradGraphFactory.drop('graphName')` should usually perform these steps for you.

**Reply by CosmoBean** [reply]
Yes, The drop didn't work and i proceeded to do the manual steps, the drop timed out

**Comment by panda0120** [comment]
I have the same problem, how to fix it?

**Reply by Zorabi** [reply]
You need to execute this before dropping the graph to ensure that there is no instance open. After deleting the graph, the Janusgraph log will always refresh table not found. You need to recreate the graph.
```
m = ${graph}.openManagement();
ids = m.getOpenInstances();
for(String id : ids){if(!id.contains("(")){m.forceCloseInstance(id)}};
m.commit();

or

m = ${graph}.openManagement();
arr = m.getOpenInstances();for(i = 0; i < arr.size();i++){ if(!arr[i].contains("current")){ m.forceCloseInstance(arr[i]) } }
m.commit()
```

**Reply by panda0120** [reply]
> You need to execute this before dropping the graph to ensure that there is no instance open. After deleting the graph, the Janusgraph log will always refresh table not found. You need to recreate the graph.在删除图表之前，您需要执行此操作以确保没有实例处于打开状态。删除图表后，Janusgraph日志将始终刷新“未找到表”信息。您需要重新创建图表。
> 
> ```
> m = ${graph}.openManagement();
> ids = m.getOpenInstances();
> for(String id : ids){if(!id.contains("(")){m.forceCloseInstance(id)}};
> m.commit();
> 
> or
> 
> m = ${graph}.openManagement();
> arr = m.getOpenInstances();for(i = 0; i < arr.size();i++){ if(!arr[i].contains("current")){ m.forceCloseInstance(arr[i]) } }
> m.commit()
> ```

Thank you very much!!!

**Reply by panda0120** [reply]
@MiracleDx but I don't want to create table again,how to stop the error log?

**Reply by Zorabi** [reply]
> @MiracleDx but I don't want to create table again,how to stop the error log?

You can try restarting your janusgraph instance

**Reply by panda0120** [reply]
@MiracleDx  this is my solution:
```
 public void dropGraph(String graphName) {
        Map<String, Object> params = new HashMap<>();
        params.put("graphName", graphName);
        this.getClient()
                .submit("m = " + graphName + ".openManagement();"
                        + "ids = m.getOpenInstances();"
                        + "for(String id : ids){if(!id.contains(\"(\")){m.forceCloseInstance(id)}};"
                        + "m.commit();"
                        + "ConfiguredGraphFactory.drop(graphName);"
                        + "ConfiguredGraphFactory.create(graphName);"
                        + "ConfiguredGraphFactory.drop(graphName);", params);
    }
```

**Reply by panda0120** [reply]
@MiracleDx This method doesn't seem to work every time.

**Reply by Zorabi** [reply]
@panda0120 But this always works for me. If the current operation hang up, you can start a new Gremlin Console to execute the following code. Once it’s done, the previous operation will be released, and the deletion will succeed.
```
m = ${graph}.openManagement();
ids = m.getOpenInstances();
for(String id : ids){if(!id.contains("(")){m.forceCloseInstance(id)}};
m.commit();
```

**Reply by panda0120** [reply]
> @panda0120 But this always works for me. If the current operation hang up, you can start a new Gremlin Console to execute the following code. Once it’s done, the previous operation will be released, and the deletion will succeed.
> 
> ```
> m = ${graph}.openManagement();
> ids = m.getOpenInstances();
> for(String id : ids){if(!id.contains("(")){m.forceCloseInstance(id)}};
> m.commit();
> ```

Now i give up to drop graph without shutdown the database, after i drop a graph i will restart all instance of my janusgraph db.

**Reply by Zorabi** [reply, marked as answer]
> > @panda0120 But this always works for me. If the current operation hang up, you can start a new Gremlin Console to execute the following code. Once it’s done, the previous operation will be released, and the deletion will succeed.
> > ```
> > m = ${graph}.openManagement();
> > ids = m.getOpenInstances();
> > for(String id : ids){if(!id.contains("(")){m.forceCloseInstance(id)}};
> > m.commit();
> > ```
> 
> Now i give up to drop graph without shutdown the database, after i drop a graph i will restart all instance of my janusgraph db.

I usually open three Gremlin windows: the first one is for dropping the graph, the second one is for executing the command to shut down the instances, and the third one is for checking whether the log ‘graph not found’ appears in the graph log.

I first execute the graph-dropping command. When it hangs, I start running the command to shut down the instances and then monitor the log. The appearance of the ‘graph not found message’ indicates that the graph has been successfully deleted.Then I will recreate the deleted graph, because the graph not found message will keep appearing in the log until the instance is restarted or the graph is recreated.

The command I currently use to shut down all instances is as follows:
```
ConfiguredGraphFactory.getGraphNames().stream().map(name -> ConfiguredGraphFactory.open(name)).map(graph -> graph.openManagement()).peek(m -> m.getOpenInstances().stream().filter(e -> !e.contains("current")).forEach(e -> m.forceCloseInstance(e))).forEach(m -> m.commit())
```

**Comment by porunov** [comment]
Ghost instances is a common problem that can be solved by:
- Force-closing ghost instances;
- (Optionally) ensuring JanusGraph instances can reuse the same instance id;

To let JanusGraph know that it's OK to reuse instance ID you can set the following properties:
```
graph.unique-instance-id=${YOUR_UNIQUE_INSTANCE_ID}
graph.replace-instance-if-exists=true
```

Closing ghost instances:
```
mgmt = graph.openManagement()
mgmt.getOpenInstances().forEach(instance -> {
   // You need to force close only ghost instances. Do not close active normal instances.
   // you need to develop your own isGhostInstance method implementation 
   if(isGhostInstance(instance)){
      mgmt.forceCloseInstance(instance);
   }
});
mgmt.commit();
```

Related: https://github.com/JanusGraph/janusgraph/discussions/3195#discussioncomment-3595012
